File: Python/QuickSort.py

def swap(l, pos1, pos2):
    temp = l[pos1]
    l[pos1] = l[pos2]
    l[pos2] = temp

def partition(l, start, end):
    pivot = l[start]
    while True:
        while l[start] < pivot:
            start = start + 1

        while l[end] > pivot:
            end = end - 1

        if start >= end:
            return end

        swap(l, start, end)

        start = start + 1
        end = end - 1

def quick_sort(l, start, end):
    if start < end:
        p = partition(l, start, end)
        quick_sort(l, start, p)
        quick_sort(l, p+1, end)

if __name__ == "__main__":
    a = [9,8,5,7,6,2,4,3,1]
    print(a)
    quick_sort(a, 0, len(a)-1)
    print(a)

File: src/cpp/QuickSort.cpp
#include <iostream>
#include <vector>

using namespace std;

int partition(vector<int> &vector, int start, int end) {
  int pivot = vector[end];
  int index = start;

  for (int i = start; i < end; i++) {
    if (vector[i] < pivot) {
      int temp = vector[i];
      vector[i] = vector[index];
      vector[index] = temp;
      index++;
    }
  }

  if (pivot <= vector[index]) {
    vector[end] = vector[index];
    vector[index] = pivot;
  }

  return index;
}

void quickSort(vector<int> &vector, int start, int end) {
  if (start < end) {
    int pivot = partition(vector, start, end);
    quickSort(vector, start, pivot - 1);
    quickSort(vector, pivot + 1, end);
  }
}

void quickSort(vector<int> &vector) { quickSort(vector, 0, vector.size() - 1); }

void showVector(vector<int> vector) {
  for (uint32_t i = 0; i < vector.size(); ++i) {
    cout << vector[i] << ", ";
  }
  cout << "\n";
}

int main() {
  vector<int> vector;
  for (uint32_t i = 0; i < 10; ++i) {
    vector.push_back(rand() % 100);
  }
  showVector(vector);
  quickSort(vector);
  showVector(vector);
}
